modules/proposals/engine/src/lib.rs

```diff
@@ -0,0 +1,348 @@
+use rstd::prelude::*;
+use runtime_primitives::traits::EnsureOrigin;
+use srml_support::{decl_event, decl_module, decl_storage, dispatch, ensure, StorageDoubleMap};
+use system::ensure_root;
+
+const DEFAULT_TITLE_MAX_LEN: u32 = 100;
+const DEFAULT_BODY_MAX_LEN: u32 = 10_000;
+
+/// Proposals engine trait.
+pub trait Trait: system::Trait + timestamp::Trait {
+    /// Engine event type.
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+
+    /// Origin from which proposals must come.
+    type ProposalOrigin: EnsureOrigin<Self::Origin, Success = Self::AccountId>;
+
+    /// Origin from which votes must come.
+    type VoteOrigin: EnsureOrigin<Self::Origin, Success = Self::AccountId>;
+
+    /// Provides data for voting. Defines maximum voters count for the proposal.
+    type TotalVotersCounter: VotersParameters;
+
+    /// Converts proposal code binary to executable representation
+    type ProposalCodeDecoder: ProposalCodeDecoder;
+}
+
+decl_event!(
+    pub enum Event<T>
+    where
+        <T as system::Trait>::AccountId
+    {
+    	/// Emits on proposal creation.
+        /// Params:
+        /// * Account id of a proposer.
+        /// * Id of a newly created proposal after it was saved in storage.
+        ProposalCreated(AccountId, u32),
+
+        /// Emits on proposal cancellation.
+        /// Params:
+        /// * Account id of a proposer.
+        /// * Id of a cancelled proposal.
+        ProposalCanceled(AccountId, u32),
+
+        /// Emits on proposal veto.
+        /// Params:
+        /// * Id of a vetoed proposal.
+        ProposalVetoed(u32),
+
+        /// Emits on proposal status change.
+        /// Params:
+        /// * Id of a updated proposal.
+        /// * New proposal status
+        ProposalStatusUpdated(u32, ProposalStatus),
+
+        /// Emits on voting for the proposal
+        /// Params:
+        /// * Voter - an account id of a voter.
+        /// * Id of a proposal.
+        /// * Kind of vote.
+        Voted(AccountId, u32, VoteKind),
+    }
+);
+
+// Storage for the proposals module
+decl_storage! {
+    trait Store for Module<T: Trait> as ProposalsEngine{
+        /// Map proposal by its id.
+        pub Proposals get(fn proposals): map u32 => Proposal<T::BlockNumber, T::AccountId>;
+
+        /// Count of all proposals that have been created.
+        pub ProposalCount get(fn proposal_count): u32;
+
+        /// Map proposal executable code by proposal id.
+        ProposalCode get(fn proposal_codes): map u32 =>  Vec<u8>;
+
+        /// Map votes by proposal id.
+        VotesByProposalId get(fn votes_by_proposal): map u32 => Vec<Vote<T::AccountId>>;
+
+        /// Ids of proposals that are open for voting (have not been finalized yet).
+        pub ActiveProposalIds get(fn active_proposal_ids): BTreeSet<u32>;
+
+        /// Proposal tally results map
+        pub(crate) TallyResults get(fn tally_results): map u32 => TallyResult<T::BlockNumber>;
```

Comment:
1. We can do that. I want to avoid bloated god-object, but we can split it later if necessary. I'll try to put tally results into the map.
2. Not sure I understand - `risky iteration in the end in #votes`

Reply:
I love God-objects :D, the compiler protects me, I can sleep well, ...with runtime invariants... I stay up at night :D

Reply:
We must always be careful to not allow iteration in the runtime, where users control the size of the iteration, as this becomes a denial of service attack, since this is not metered environment, like EVM gas model. The users can make #voters very large, we cannot allow this.

Reply:
Conclusion: the point aobut #votes is not correct, as you pointed out, because of the context of our our voting scheme works.